README.md

[![Arduino CI](https://github.com/RobTillaart/SGP30/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/SGP30/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/SGP30/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/SGP30/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/SGP30/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/SGP30/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/SGP30.svg?maxAge=3600)](https://github.com/RobTillaart/SGP30/releases)


# SGP30

Arduino library for SGP30 environment sensor

Warning: experimental, library is not functional complete yet.


## Description

The SGP30 from Sensirion is an environment sensor that measures H2 and Ethanol in the air. 
From these numbers an intern algorithm in the sensor derives an CO2 equivalent and a TVOC 
measurement. The library has an experimental conversion for H2 and Ethanol.

The CO2 units are ppm, the TVOC units are ppb. Units for H2 and Ethanol are ppm. 
Note that for larger concentrations the resolution of the measurements drops, see datasheet.

The library supports 2 types of interfaces, a synchronous and an asynchronous interface. 
The sync interface is blocking for up to 40 milliseconds which was enough to trigger the 
implementation of an async interface. 

Note: the sync interface is implemented with the async interface.


#### Sample frequency

The CO2 and TVOC values can be read up to once per second (1 Hz). 
Ethanol and H2, the raw data can be sampled up to 40 Hz.

The first 15 seconds the sensor needs to stabilize. Thereafter one gets real data.


#### I2C performance

The SGP30 works with I2C bus at 100 KHz and 400 KHz. In a short test it worked well up to 500 KHz. 
A faster I2C clock does not give the sync interface much (relative) gain, 
however for the async the relative gain is much more.

(indicative test run with UNO - IDE 1.8.13, CO2 and TVOC only, times in micros)

|  I2C speed | measure() | request() | read() |
|:----------:|:---------:|:---------:|:------:|
|  100 kHz   |   12360   |    336    |   732  |
|  200 kHz   |   12212   |    196    |   408  |
|  300 kHz   |   12168   |    144    |   300  |
|  400 kHz   |   12140   |    132    |   264  |
|  500 kHz   |   12128   |    124    |   236  |

Note the blocking of measure().


#### Multiple sensors.

The SGP30 sensor has a fixed I2C address so only one sensor per I2C bus can be used. 
If one needs more, one should use an I2C multiplexer or an MCU with multiple I2C buses 
or switch the VCC as a sort of ChipSelect signal.


## Interface

### Constructor

- **SGP30(TwoWire \*wire = &Wire)** Constructor with optional the Wire interface as parameter.
- **bool begin()** .
- **bool begin(uint8_t sda, uint8_t scl)** idem, for the ESP32 where one can choose the I2C pins.
- **bool isConnected()** checks if the address is visible on the I2C bus.


### Meta

- **bool getID()** reads the sensor ID into 12 bytes. (needs rework).
- **uint16_t getFeatureSet()** returns 0x0022, indicates that commands used in this library are supported.
- **bool measureTest()** verify the chip is working.


### Synchronous measurements

- **uint32_t lastMeasurement()** timestamp in milliseconds of the last sync measurement made. This convenience function is useful to prevent reading the sensor too often.
- **bool measure(bool all = false)** if all == false, only the TVOC and CO2 are updated (slow due to blocking), if all == true, also the H2 and the Ethanol values are updated (even slower). Note the measurement is slow as there is an active blocking until the sensor is done. If the last measurement is less than a second ago, no measurement is made and the function returns false.


### A-synchronous measurements

With the async interface, the user should control that reads are at least one second apart. The user should also take care not to mix up different requests. See examples. 

- **void request()** sends a request to the sensor to read CO2 and TVOC. 
- **bool read()** returns true if the last request is more than 12 milliseconds ago the CO2 and TVOC are read and updated. Otherwise false is returned.
- **void requestRaw()** sends a request to the sensor to read H2 and Ethanol.
- **bool readRaw()** returns true if the last request is more than 25 milliseconds ago the H2 and Ethanol are read and updated. Otherwise false is returned.


### Get the data

The library caches the last read values, and these are the functions to access them. 

- **uint16_t getTVOC()** gets the TVOC concentration (ppb)
- **uint16_t getCO2()** gets the CO2 **equivalent** concentration (ppm)
- **uint16_t getH2_raw()** gets the raw H2. Units unknown.
- **uint16_t getEthanol_raw()** gets the raw Ethanol. Units unknown.


### Calibration

Check the datasheet for operating range, figure 7.

- **float setRelHumidity(float T, float RH)** sets the compensation for temperature (5-55°C) and **relative** humidity (10-95%). These values can be obtained e.g. from an SHT30, DHT22 or similar sensor. The function returns the absolute humidity. 
- **void setAbsHumidity(float absoluteHumidity)** sets the compensation for **absolute** humidity. Concentration is in gram per cubic meter (g/m3)


### Baseline functions

The baseline functions give the sensor a reference value. After running in a known condition e.g. outside in open air, one can get the baseline values as a sort of calibration. Please read the datasheet carefully before using these functions.

Note: if the sensor has no reads done, these values tend to go to zero. This is because the baselines are based upon recent reads.

- **bool getBaseline(uint16_t \*CO2, uint16_t \*TVOC)** retrieves the baseline values from the sensor.
- **void setBaseline(uint16_t CO2, uint16_t TVOC)** sets the baseline values.


For faster accurate results for the TVOC under bad air conditions, read **Inceptive Baseline for TVOC measurements**
(not tested)
- **bool getTVOCBaseline(uint16_t \*TVOC)** retrieves the TVOC start value from the sensor.
- **void setTVOCBaseline(uint16_t TVOC)** sets the TVOC start value.


### Misc

- **int lastError()** returns last error. (needs rework)


### Experimental

use at own risk.

Since 0.1.2 the library has experimental support for H2 and Ethanol concentration in ppm. 

One should use these functions more as a relative indication than as an absolute measurement as it is definitely not calibrated.
Runs with different temperature and humidity (different days) give very different values including overflow and infinity.

- **float getH2()** gets the H2 concentration. Units ppm.
- **float getEthanol()** gets the Ethanol concentration. Units ppm.

The used references are based upon (1) averaging raw data in outside air at 22°C @ 1 meter and (2) the assumption that this is 0.4 resp 0.5 ppm. (Note only 1 significant digit) as mentioned in datasheet P2.

- **void  setSrefH2(uint16_t s = 13119)** // 13119 is my measurement.
- **uint16_t getSrefH2()** returns value set.
- **void setSrefEthanol(uint16_t s = 18472)** // 18472 is my measurement
- **uint16_t getSrefEthanol()** returns value set


## Todo

- redo **getID()**
- test test test ....


The CRC checking + error handling (since 0.1.4) adds around 330 bytes PROGMEM on an UNO.
There might be a need for a minimal class that only reads CO2 and TVOC, no baselines etc.
for the smallest platforms. 


## Operational

See examples


## Links

https://www.adafruit.com/product/3709 - the sensor.



CHANGELOG.md
# Change Log SGP30

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.0] - 2023-01-27
- merge PR #11 (kudos to countrysidecowboy)
- update GitHub actions
- update license 2023
- update readme.md
- minor edits


## [0.1.6] - 2022-10-28
- Add RP2040 support to build-CI.
- Add CHANGELOG.md

## [0.1.5] - 2021-12-28  
- update library,json
- update readme
- update license
- minor edits

## [0.1.4] - 2021-07-01  
- add CRC checking

## [0.1.3] - 2021-06-26 
- add get/setTVOCbaseline()

## [0.1.2] - 2021-06-26
- experimental add units  H2 + Ethanol

## [0.1.1] - 2021-06-26
- add get/setBaseline ++

## [0.1.0] - 2021-06-24
- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2021-01-07
// PURPOSE: unit tests for the SET library
//          https://github.com/RobTillaart/SET
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)


#include <ArduinoUnitTests.h>


#include "SGP30.h"



unittest_setup()
{
  fprintf(stderr, "SGP30_LIB_VERSION: %s\n", (char *) SGP30_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constructor)
{
  SGP30 SGP;

  assertTrue(SGP.begin());
  assertTrue(SGP.isConnected());

  assertEqual(0x00, SGP.lastMeasurement());
  assertEqual(0x00, SGP.lastError());
}


unittest(test_constants)
{
  assertEqual(SGP30_OK,        0x00);
  assertEqual(SGP30_ERROR_CRC, 0xFF);
  assertEqual(SGP30_ERROR_I2C, 0xFE);
}


unittest(test_defaults_core)
{
  SGP30 SGP;

  assertEqual(0x00, SGP.getCO2());
  assertEqual(0x00, SGP.getTVOC());
  assertEqual(0x00, SGP.getH2_raw());
  assertEqual(0x00, SGP.getEthanol_raw());
  assertEqualFloat(6.71302e+10, SGP.getH2(), 1e7);
  assertEqualFloat(1.86462e+15, SGP.getEthanol(), 1e12);
}


unittest(test_sref_H2)
{
  SGP30 SGP;

  assertEqual(13119, SGP.getSrefH2());

  SGP.setSrefH2(0);
  assertEqual(0, SGP.getSrefH2());

  SGP.setSrefH2(10000);
  assertEqual(10000, SGP.getSrefH2());

  SGP.setSrefH2();
  assertEqual(13119, SGP.getSrefH2());
}


unittest(test_sref_Ethanol)
{
  SGP30 SGP;

  assertEqual(18472, SGP.getSrefEthanol());

  SGP.setSrefEthanol(0);
  assertEqual(0, SGP.getSrefEthanol());

  SGP.setSrefEthanol(10000);
  assertEqual(10000, SGP.getSrefEthanol());

  SGP.setSrefEthanol();
  assertEqual(18472, SGP.getSrefEthanol());
}


unittest_main()


//  -- END OF FILE --


SGP30.cpp
//
//    FILE: SGP30.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.2.0
//    DATE: 2021-06-24
// PURPOSE: SGP30 library for Arduino
//     URL: https://github.com/RobTillaart/SGP30
//          https://www.adafruit.com/product/3709


#include "SGP30.h"


/////////////////////////////////////////////////////
//
//  CONSTRUCTOR
//
SGP30::SGP30(TwoWire *wire)
{
  _address  = 0x58;
  _wire     = wire;

  _tvoc     = 0;
  _co2      = 0;
  _h2       = 0;
  _ethanol  = 0;

  _lastTime = 0;
  _error    = SGP30_OK;
}


#if defined (ESP8266) || defined(ESP32)
bool SGP30::begin(uint8_t dataPin, uint8_t clockPin)
{
  _wire      = &Wire;
  if ((dataPin < 255) && (clockPin < 255))
  {
    _wire->begin(dataPin, clockPin);
  } else {
    _wire->begin();
  }
  if (! isConnected()) return false;
  return true;
}
#endif


bool SGP30::begin()
{
  _wire->begin();
  if (! isConnected()) return false;
  _init();
  return true;
}


bool SGP30::isConnected()
{
  _wire->beginTransmission(_address);
  return ( _wire->endTransmission() == 0);
}


//  INITIAL VERSION - needs optimization
bool SGP30::getID()
{
  _command(0x3682);
  delay(1);
  if (_wire->requestFrom(_address, (uint8_t)9) == 9)
  {
    for (uint8_t i = 0, j = 0; i < 3; i++)
    {
      _id[j++] = _wire->read();
      _id[j++] = _wire->read();
      uint8_t crc = _wire->read();
      uint16_t val = _id[j-2] * 256 + _id[j-1];
      if (_CRC8(val) != crc)
      {
        _error = SGP30_ERROR_CRC;
        return false;
      }
    }
    _error = SGP30_OK;
    return true;
  }
  _error = SGP30_ERROR_I2C;
  return false;
}


//  expect to return 0x0022
uint16_t SGP30::getFeatureSet()
{
  _command(0x202F);
  delay(10);
  uint16_t rv = 0;
  if (_wire->requestFrom(_address, (uint8_t)3) == 3)
  {
    rv = _wire->read() << 8;
    rv  += _wire->read();
    uint8_t crc = _wire->read();
    if (_CRC8(rv) != crc)
    {
      _error = SGP30_ERROR_CRC;
      return rv;
    }
    _error = SGP30_OK;
    return rv;
  }
  _error = SGP30_ERROR_I2C;
  return 0xFFFF;
}


//  WARNING: resets all devices on I2C bus.
void SGP30::GenericReset()
{
  _command(0x0006);
  _init();
};


bool SGP30::measureTest()
{
  uint16_t rv = 0;
  _command(0x2032);
  delay(220);             //  Page 11
  if (_wire->requestFrom(_address, (uint8_t)3) == 3)
  {
    rv = _wire->read() << 8;
    rv  += _wire->read();
    uint8_t crc = _wire->read();
    if (_CRC8(rv) != crc)
    {
      _error = SGP30_ERROR_CRC;
      return false;
    }
    _error = SGP30_OK;
    return rv == 0xD400;
  }
  _error = SGP30_ERROR_I2C;
  return false;
}


/////////////////////////////////////////////////////
//
//  MEASUREMENT
//

//  SYNCHRONUOUS MODUS
bool SGP30::measure(bool all)      //  this is the workhorse
{
  //  1 second between measurements. Page 9
  if (millis() - _lastTime < 1000) return false;
  _lastTime = millis();

  request();
  delay(12);  //  blocking!!
  read();

  if (not all) return true;

  requestRaw();
  delay(25);  //  blocking!!
  readRaw();
  return true;
}


//  A-SYNCHRONUOUS INTERFACE
void SGP30::request()
{
  _lastRequest = millis();
  _command(0x2008);
}


bool SGP30::read()
{
  if (_lastRequest == 0) return false;
  if (millis() - _lastRequest < 13) return false;  //  Page 11
  _lastRequest = 0;

  if (_wire->requestFrom(_address, (uint8_t)6) != 6)
  {
    _error = SGP30_ERROR_I2C;
    return false;
  }
  _co2  =  _wire->read() << 8;
  _co2  += _wire->read();
  uint8_t crc = _wire->read();
  if (_CRC8(_co2) != crc)
  {
    _error = SGP30_ERROR_CRC;
    return false;
  }
  _tvoc =  _wire->read() << 8;
  _tvoc += _wire->read();
  crc = _wire->read();
  if (_CRC8(_tvoc) != crc)
  {
    _error = SGP30_ERROR_CRC;
    return false;
  }
  _error = SGP30_OK;
  return true;
}


void SGP30::requestRaw()
{
  _lastRequest = millis();
  _command(0x2050);
}


bool SGP30::readRaw()
{
  if (_lastRequest == 0) return false;
  if (millis() - _lastRequest < 26) return false;  //  Page 11
  _lastRequest = 0;

  if (_wire->requestFrom(_address, (uint8_t)6) != 6)
  {
    _error = SGP30_ERROR_I2C;
    return false;
  }
  _h2      =  _wire->read() << 8;
  _h2      += _wire->read();
  uint8_t crc = _wire->read();
  if (_CRC8(_h2) != crc)
  {
    _error = SGP30_ERROR_CRC;
    return false;
  }
  _ethanol =  _wire->read() << 8;
  _ethanol += _wire->read();
  crc = _wire->read();
  if (_CRC8(_ethanol) != crc)
  {
    _error = SGP30_ERROR_CRC;
    return false;
  }
  _error = SGP30_OK;
  return true;
}


//  experimental - datasheet Page 2
//  1.953125e-3 = 1/512
float SGP30::getH2()
{
  float cref = 0.5;  //  ppm
  return cref * exp((_srefH2 - _h2) * 1.953125e-3);
}


float SGP30::getEthanol()
{
  float cref = 0.4;  //  ppm
  return cref * exp((_srefEth - _ethanol) * 1.953125e-3);
}


/////////////////////////////////////////////////////
//
//  CALIBRATION
//

//  T  in °C
//  RH == RelativeHumidity
float SGP30::setRelHumidity(float T, float RH)      //  Page 10
{
  //  page 10 datasheet
  //  AH = AbsoluteHumidity
  //  uint16_t AH = 216.7 * RH/100 * 6.117 * exp((17.62 * T)/(243.12 + T)) / (273.15 + T);
  float absoluteHumidity = (2.167 * 6.112) * RH ;
  absoluteHumidity *= exp((17.62 * T)/(243.12 + T));
  absoluteHumidity /=  (273.15 + T);

  setAbsHumidity(absoluteHumidity);
  return absoluteHumidity;
}


void SGP30::setAbsHumidity(float absoluteHumidity)
{
  uint16_t AH = absoluteHumidity;
  uint8_t tmp = (absoluteHumidity - AH) * 256;
  AH = (AH << 8) | tmp;

  _command(0x2061, AH);     //  Page 11
}


void SGP30::setBaseline(uint16_t CO2, uint16_t TVOC)
{
  _command(0x201E, TVOC, CO2);
}


bool SGP30::getBaseline(uint16_t *CO2, uint16_t *TVOC)
{
  _command(0x2015);
  if (_wire->requestFrom(_address, (uint8_t)6) != 6)
  {
    _error = SGP30_ERROR_I2C;
    return false;
  }
  *CO2  =  _wire->read() << 8;
  *CO2  += _wire->read();
  uint8_t crc = _wire->read();
  if (_CRC8(*CO2) != crc)
  {
    _error = SGP30_ERROR_CRC;
    return false;
  }
  *TVOC =  _wire->read() << 8;
  *TVOC += _wire->read();
  crc = _wire->read();
  if (_CRC8(*TVOC) != crc)
  {
    _error = SGP30_ERROR_CRC;
    return false;
  }
  _error = SGP30_OK;
  return true;
}


void SGP30::setTVOCBaseline(uint16_t TVOC)
{
  _command(0x2077, TVOC);
}


bool SGP30::getTVOCBaseline(uint16_t *TVOC)
{
  _command(0x20B3);
  if (_wire->requestFrom(_address, (uint8_t)3) != 3)
  {
    _error = SGP30_ERROR_I2C;
    return false;
  }
  *TVOC =  _wire->read() << 8;
  *TVOC += _wire->read();
  uint8_t crc = _wire->read();
  if (_CRC8(*TVOC) != crc)
  {
    _error = SGP30_ERROR_CRC;
    return false;
  }
  _error = SGP30_OK;
  return true;
}


/////////////////////////////////////////////////////
//
//  MISCELLANEOUS
//
int SGP30::lastError()
{
  int error = _error;
  _error = SGP30_OK;
  return error;
}


/////////////////////////////////////////////////////
//
//  PRIVATE
//
int SGP30::_command(uint16_t cmd)
{
  _wire->beginTransmission(_address);
  _wire->write(cmd >> 8);
  _wire->write(cmd & 0xFF);
  _error = _wire->endTransmission();
  return _error;
}


int SGP30::_command(uint16_t cmd, uint16_t v1)
{
  _wire->beginTransmission(_address);
  _wire->write(cmd >> 8);
  _wire->write(cmd & 0xFF);
  _wire->write(v1 >> 8);
  _wire->write(v1 & 0xFF);
  _wire->write(_CRC8(v1));
  _error = _wire->endTransmission();
  return _error;
}


int SGP30::_command(uint16_t cmd, uint16_t v1, uint16_t v2)
{
  _wire->beginTransmission(_address);
  _wire->write(cmd >> 8);
  _wire->write(cmd & 0xFF);
  _wire->write(v1 >> 8);
  _wire->write(v1 & 0xFF);
  _wire->write(_CRC8(v1));
  _wire->write(v2 >> 8);
  _wire->write(v2 & 0xFF);
  _wire->write(_CRC8(v2));
  _error = _wire->endTransmission();
  return _error;
}


//  for sending command - CRC lib.
//  always 2 bytes
uint8_t SGP30::_CRC8(uint16_t data)
{
  uint8_t val[2];
  val[0] = data >> 8;
  val[1] = data & 0xFF;

  uint8_t crc = 0xFF;             //  start value
  for(uint8_t i = 0; i < 2; i++)
  {
    crc ^= val[i];
    for (uint8_t b = 8; b > 0; b--)
    {
      if (crc & 0x80)
        crc = (crc << 1) ^ 0x31;  //  polynomial
      else
        crc <<= 1;
    }
  }
  return crc;
};


void SGP30::_init()
{
  _command(0x2003);
};


//  -- END OF FILE --



SGP30.h
#pragma once
//
//    FILE: SGP30.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.2.0
//    DATE: 2021-06-24
// PURPOSE: SGP30 library for Arduino
//     URL: https://github.com/RobTillaart/SGP30
//          https://www.adafruit.com/product/3709


#include "Arduino.h"
#include "Wire.h"

#define SGP30_LIB_VERSION               (F("0.2.0"))

#define SGP30_OK                        0x00
#define SGP30_ERROR_CRC                 0xFF
#define SGP30_ERROR_I2C                 0xFE


class SGP30
{
public:
  explicit SGP30(TwoWire *wire = &Wire);

#if defined (ESP8266) || defined(ESP32)
  bool     begin(uint8_t sda, uint8_t scl);
#endif
  bool     begin();
  bool     isConnected();
  //  WARNING resets all I2C devices that support this call !!
  void     GenericReset();


  //  META
  bool     getID();
  uint16_t getFeatureSet();
  bool     measureTest();   //  need a better name


  //  MEASUREMENTS
  //  lastMeasurement works only with measure();
  //                 not with async calls.
  uint32_t lastMeasurement() { return _lastTime; };
  //  all == false ==> only TVOC and eCO2
  //  blocks 15 - 40 milliseconds  (depends on parameter all).
  bool     measure(bool all = false);

  //  async interface
  void     request();
  bool     read();
  void     requestRaw();
  bool     readRaw();

  //  get the data                                 //   UNITS
  uint16_t getTVOC()        { return _tvoc; };     //   PPB
  uint16_t getCO2()         { return _co2; };      //   PPM
  uint16_t getH2_raw()      { return _h2; };       //   UNKNOWN
  uint16_t getEthanol_raw() { return _ethanol; };  //   UNKNOWN
  float    getH2();         //  experimental       //   PPM
  float    getEthanol();    //  experimental       //   PPM


  //  CALIBRATION - read datasheet
  //  T  in °C
  //  RH == 0..100
  float    setRelHumidity(float T, float RH);      //  P10
  //  Absolute humidity in grams / m3
  //  set Abs Hum to 0 to disables it...
  void     setAbsHumidity(float absoluteHumidity);

  void     setBaseline(uint16_t CO2, uint16_t TVOC);
  bool     getBaseline(uint16_t *CO2, uint16_t *TVOC);

  //  See Inceptive Baseline for TVOC measurements in data sheet
  void     setTVOCBaseline(uint16_t TVOC);
  bool     getTVOCBaseline(uint16_t *TVOC);


  //  EXPERIMENTAL
  //  13119 = average raw measured outside 22°C  (example)
  void      setSrefH2(uint16_t s = 13119)      { _srefH2 = s; };
  uint16_t  getSrefH2()                        { return _srefH2; };
  //  18472 = average raw measured outside 22°C  (example)
  void      setSrefEthanol(uint16_t s = 18472) { _srefEth = s; };
  uint16_t  getSrefEthanol()                   { return _srefEth; };


  //  MISCELLANEOUS
  int      lastError();

  //  need an access function to prevent writing _id[]
  //  for now it is OK.  TODO private?
  uint8_t  _id[6];


private:
  uint8_t  _address;
  int      _error;
  uint32_t _lastTime    = 0;
  uint32_t _lastRequest = 0;

  //  TODO improve?
  int      _command(uint16_t cmd);
  int      _command(uint16_t cmd, uint16_t v1);
  int      _command(uint16_t cmd, uint16_t v1, uint16_t v2);
  uint8_t  _CRC8(uint16_t val);
  void     _init();

  uint16_t _tvoc;
  uint16_t _co2;
  uint16_t _h2;
  uint16_t _ethanol;

  //  experimental
  //  average raw values measured outside 22°C
  uint16_t _srefH2  = 13119;
  uint16_t _srefEth = 18472;

  TwoWire*  _wire;
};


//  -- END OF FILE --

